### PR TITLE
[feature] #3871: `Register<AssetDefinition>` permissions

### LINKED ISSUE #3871: `Register<AssetDefinition>` permissions
ATM, there are no permissions for registering asset definitions in the default validator, i.e. all `Register<AssetDefinition>` operations are never denied. For example it is allowed that `alice@wonderland` registers a new asset definition `xor#genesis`. Should we disallow `alice@wonderland` from accessing other domains by default and introduce some tokens to grant such a privilege?

### PR DESCRIPTION
**Pull request by ilchu**
## Description

With this change, having a permission token is required to register `AssetDefinition`s in domains for accounts besides the domain owner.

* added `CanMintUserAsset` permission token
* added `CanRegisterAccountInDomain` permission token
* added `CanRegisterAssetDefinitionInDomain` permission_token
* renamed `CanXXXAssetsWithDefinition` to `CanXXXAssetWithDefinition`
* renamed `Grant/Revoke::permission_token` to `Grant/Revoke::permission`

### Linked issue

Closes #3871. 

### Benefits
More granular control over roles in permissioned blockchain.

### Checklist

- [x] I've read `CONTRIBUTING.md`
- [x] I've used the standard signed-off commit format (or will squash just before merging)
- [x] All applicable CI checks pass (or I promised to make them pass later)
- [x] I've adapted an integration test for the code changes
- [x] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: smart_contract/executor/src/default.rs [hunk 1]
```diff
@@ -563,10 +564,42 @@ pub mod account {
...
+            pass!(executor);
+        }
+
+        deny!(
+            executor,
+            "New `AssetDefinition` can only be registered by domain owner or if granted `CanRegisterAssetDefinitionInDomain` token"
+        );
+    }
+
     pub fn visit_unregister_asset_definition<V: Validate + ?Sized>(
```
Comment (Arjentix) [inline smart_contract/executor/src/default.rs:603]: I think this functions should have a check if `authority` owns the domain

Comment (Erigara) [reply smart_contract/executor/src/default.rs:603]: AFAIK this check incorporated in `is_asset_definition_owner`

**Review by Erigara — DISMISSED** [review]
lgtm

File: smart_contract/executor/derive/src/default.rs [hunk 2]
```diff
@@ -138,6 +138,7 @@ pub fn impl_derive_visit(emitter: &mut Emitter, input: &syn2::DeriveInput) -> To
         "fn visit_transfer_asset(operation: Transfer<Asset, NumericValue, Account>)",
         "fn visit_set_asset_key_value(operation: SetKeyValue<Asset>)",
         "fn visit_remove_asset_key_value(operation: RemoveKeyValue<Asset>)",
+        "fn visit_register_asset_definition(operation: Register<AssetDefinition>)",
```
[the remark this answers is not in the document]
Comment (ilchu) [reply smart_contract/executor/derive/src/default.rs:141]: I thought I deleted `visit_register_domain` function from that custom executor in my previous PR for the same reason. But I'll make sure to double check, yes. Also this seems to me the biggest pain point, having these two parts are disconnected from each other, so it's easy to forget to make the change in the second place after the first. I should at least leave a note in the `executor/default.rs` module.

[the remark this answers is not in the document]
Comment (Arjentix) [reply smart_contract/executor/derive/src/default.rs:141]: @ilchu , how is it going? I can see no new changes in this file

**Comment by Arjentix** [comment]
Also there is an unrelated commit in this PR. You should probably remove it

File: smart_contract/executor/derive/src/default.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ilchu) [reply smart_contract/executor/derive/src/default.rs:142]: Totally slipped me, now that TODO's fixed.

File: client/tests/integration/smartcontracts/create_nft_for_every_user_trigger/src/lib.rs [hunk 3]
```diff
@@ -21,9 +21,21 @@ fn main(_owner: AccountId, _event: Event) {
...
+        "genesis".parse().dbg_unwrap(),
+        "garden_of_live_flowers".parse().dbg_unwrap(),
+    ];
+
     for account in accounts_cursor {
         let account = account.dbg_unwrap();
 
+        if bad_domain_ids
+            .iter()
+            .any(|bad_id| account.id().domain_id() == bad_id)
```
Comment (Arjentix) [inline client/tests/integration/smartcontracts/create_nft_for_every_user_trigger/src/lib.rs:34]:
```suggestion
        if bad_domain_ids.contains(account.id().domain_id())
```

File: client/tests/integration/smartcontracts/create_nft_for_every_user_trigger/src/lib.rs [hunk 4]
```diff
@@ -43,6 +55,7 @@ fn main(_owner: AccountId, _event: Event) {
         let account_nft_id = AssetId::new(nft_id, account.id().clone());
         let account_nft = Asset::new(account_nft_id, Metadata::new());
 
+        iroha_trigger::log::warn!(&format!("TRYING TO REGISTER {nft_definition}"));
```
Comment (Arjentix) [inline client/tests/integration/smartcontracts/create_nft_for_every_user_trigger/src/lib.rs:58]: Should be removed or `info` level at least

**Comment by ilchu** [comment]
> Looks like doc test is failing because it uses feature-gated `dbg`

I'll just ignore that doctest because it fails with all features flag even with config directive included. Or am I missing something here?

File: smart_contract/src/lib.rs [hunk 5]
```diff
@@ -47,7 +47,7 @@ unsafe extern "C" fn _iroha_smart_contract_dealloc(offset: *mut u8, len: usize)
 ///
 /// # Example
 ///
-/// ```
+/// ```ignore
```
[the remark this answers is not in the document]
Comment (ilchu) [reply smart_contract/src/lib.rs:50]: I believe in CI it runs with `--all-features` and it fails in that case even locally for me in that case with the following: 

```
ld: warning: no platform load command found in '/private/var/folders/rs/3t6nkp7j1cq7fpvrrxr19qpw0000gn/T/rustcaxtnFu/symbols.o', assuming: macOS
          ld: Undefined symbols:
            _dbg, referenced from:
                iroha_smart_contract_utils::debug::dbg::ha2c8a6701aaf000e in libiroha_smart_contract-faa30d5efbc2e314.rlib[10](iroha_smart_contract-faa30d5efbc2e314.21789y9go0cgeteo.rcgu.o)
```

[the remark this answers is not in the document]
Comment (mversic) [reply smart_contract/src/lib.rs:50]: ok, it seems that doctests are not built with either `cfg(test)` or `cfg(doctest)`. [Here](https://github.com/rust-lang/rust/issues/67295) is relevant issue. 

I don't think there is much we can do

**Review by mversic — APPROVED** [review]

**Review by Arjentix — APPROVED** [review]

**Review by Arjentix — DISMISSED** [review]

**Review by mversic — APPROVED** [review]

**Review by mversic — DISMISSED** [review]

**Review by mversic — APPROVED** [review]

File: core/src/smartcontracts/isi/account.rs [hunk 6]
```diff
@@ -198,14 +198,28 @@ pub mod isi {
...
+                asset_definition.owned_by = destination_id.clone();
+
+                wsv.emit_events(Some(AssetDefinitionEvent::OwnerChanged(
+                    AssetDefinitionOwnerChanged {
+                        asset_definition_id: object,
+                        new_owner: destination_id,
+                    },
+                )));
+            } else {
+                // TODO: Return some error
```
Comment (Arjentix) [inline core/src/smartcontracts/isi/account.rs:221]: Should be addressed I gues

File: core/src/smartcontracts/isi/domain.rs [hunk 7]
```diff
@@ -287,12 +287,26 @@ pub mod isi {
...
+
+            if domain.owned_by == source_id {
+                domain.owned_by = destination_id.clone();
+
+                wsv.emit_events(Some(DomainEvent::OwnerChanged(DomainOwnerChanged {
+                    domain_id: object,
+                    new_owner: destination_id,
+                })));
+            } else {
+                // TODO: Return some error
```
Comment (Arjentix) [inline core/src/smartcontracts/isi/domain.rs:308]: Same

**Review by Arjentix — CHANGES_REQUESTED** [review]

File: core/src/smartcontracts/isi/account.rs (hunk 6 above)
Comment (mversic) [reply core/src/smartcontracts/isi/account.rs]: ok, I addressed it

**Review by mversic — APPROVED** [review]

**Review by Arjentix — APPROVED** [review]

**Pull request merged**